---
title: "General overlap in features affected by mutations and switches"
author: "HÃ©ctor Climente"
date: "10/09/2015"
output:
  html_document:
    toc: true
---

```{r environment, include=FALSE }
source("~/SmartAS/Pipeline/scripts/variablesAndFunctions.r")
library(knitr)
library(DT)
results_filepath <- paste0(workingDir,"/mutations/")
wd <- paste0("~/SmartAS/Pipeline/notebook/20151009_domain_ranking_files/data/")
```

```{r get_files, include=FALSE }
dir.create(wd,recursive=T)

file.copy(paste0(results_filepath,'feature_enrichment.txt'),wd,overwrite = T)
file.copy(paste0(workingDir,"/Data/Databases/prosite2go.clean.txt"),wd,overwrite = T)
file.copy(paste0(workingDir,"/Data/Databases/Pfam2go.clean.txt"),wd,overwrite = T)
```

We look for structural features that are frequently mutated and we compare them to the structural features that are usually switched. The interest is to look for splicing changes that are equivalent to mutation changes. We use the following parameters:

* Mutation frequency: proportion of all the functional mutations that affect one particular feature. Measured taking into account only the features in the most abundant isoform of each gene.

* Switch frequency: sum of all the patients that, per switch, have a feature affected.

* Domain frequency: frequency of a feature in the proteome, counting the features in the most abundant isoform for each gene. Used to normalize the previous quantities.

We annotated the domains relatively highly mutated and highly switched (highly meaning > quantile 95).

# Correlation between the frequency that mutations and switches affect each feature

## Analysis by tumor type
```{r tumor_mutVsSwitch}
feat_enrich <- read.delim(paste0(wd,'feature_enrichment.txt'), header=TRUE)
feat_enrich$MutFreq <- feat_enrich$MutIn/feat_enrich$AllMuts
feat_enrich$SwitchFreq <- feat_enrich$SwitchesIn/feat_enrich$AllSwitches
feat_enrich$DomainFrequency <- feat_enrich$DomainCount/feat_enrich$AllDomains
feat_enrich <- feat_enrich[feat_enrich$DomainFrequency != 0,]

for (cancer in cancerTypes){
  
  cat(paste0(cancer,"\n"))
  cancer.feat_enrich <- feat_enrich[feat_enrich$Cancer==cancer,]
  
  # plot normalized mutation frequency and switched frequency
  minMut <- quantile(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,0.95)
  minSwitch <- quantile(cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency,0.95)
  
  # plot raw data
  r <- cor(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,
           cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency)
  
  p <- ggplot(cancer.feat_enrich,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
    geom_point() + 
    geom_point(data=subset(cancer.feat_enrich, MutFreq/DomainFrequency > minMut & SwitchFreq/DomainFrequency > minSwitch),aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency,color=Domain)) + 
    smartas_theme() + 
    geom_smooth(method=lm) + 
    geom_text(x=20,y=100,label=paste0("R = ",round(r,2)))
  p <- direct.label(p)
  
  print(p)  
}
```

These plots leads to think that usually changes achieved by splicing are not achieved by mutations and viceversa. Still, there are some outlier domains that are a bit outside the L-shape, suggesting that some features can either be affected by any mechanism. To find out about those, we remove the features that have 0 on either SwitchFreq or MutFreq.

```{r tumor_rmOutliers_mutVsSwitch}
# plot log representations and remove those that have 0 on either side
for (cancer in cancerTypes){
  
  cat(paste0(cancer,"\n"))
  cancer.feat_enrich <- feat_enrich[feat_enrich$Cancer==cancer,]
  
  # plot normalized mutation frequency and switched frequency
  minMut <- quantile(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,0.95)
  minSwitch <- quantile(cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency,0.95)

  # filter data and get correlation
  r.df <- subset(cancer.feat_enrich, MutFreq != 0 & SwitchFreq != 0)
  r <- cor(r.df$MutFreq/r.df$DomainFrequency,r.df$SwitchFreq/r.df$DomainFrequency)
  
  p <- ggplot(r.df,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
    geom_point() + 
    geom_point(data=subset(r.df, MutFreq/DomainFrequency > minMut & SwitchFreq/DomainFrequency > minSwitch),aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency,color=Domain)) + 
    smartas_theme() + 
    geom_smooth(method=lm) + 
    geom_text(x=6,y=20,label=paste0("R = ",round(r,2)))
  p <- direct.label(p)
  
  print(p)
  
}
```

Some degree of correlation appears, particularly in the kidney tumors (kich,kirc,kirp). 

## Analysis aggregating all tumors

We apply the same process to aggregated data from all tumors.

```{r aggregated_mutVsSwitch}
feat_enrich_agg <- ddply(feat_enrich,.(Domain), summarise, 
                        MutIn=sum(MutIn), AllMuts=sum(AllMuts),
                        SwitchesIn=sum(SwitchesIn), AllSwitches=sum(AllSwitches),
                        DomainCount=sum(DomainCount),AllDomains=sum(AllDomains))
feat_enrich_agg$MutFreq <- feat_enrich_agg$MutIn/feat_enrich_agg$AllMuts
feat_enrich_agg$SwitchFreq <- feat_enrich_agg$SwitchesIn/feat_enrich_agg$AllSwitches
feat_enrich_agg$DomainFrequency <- feat_enrich_agg$DomainCount/feat_enrich_agg$AllDomains

minMut <- quantile(feat_enrich_agg$MutFreq/feat_enrich_agg$DomainFrequency,0.95)
minSwitch <- quantile(feat_enrich_agg$SwitchFreq/feat_enrich_agg$DomainFrequency,0.95)

# plot original data
r <- cor(feat_enrich_agg$MutFreq/feat_enrich_agg$DomainFrequency,
         feat_enrich_agg$SwitchFreq/feat_enrich_agg$DomainFrequency)

p <- ggplot(data=feat_enrich_agg,
            aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
  geom_point() + 
  geom_point(data=subset(feat_enrich_agg,
                         MutFreq/DomainFrequency > minMut & 
                         SwitchFreq/DomainFrequency > minSwitch),
             aes(MutFreq/DomainFrequency,
                 SwitchFreq/DomainFrequency,
                 color=Domain)) + 
  smartas_theme() + 
  geom_smooth(method=lm) + 
  geom_text(x=20,y=100,label=paste0("R = ",round(r,2)))
p <- direct.label(p)

p
```

```{r aggregated_log_mutVsSwitch }

# plot log2 representations and remove those that have 0 on either side
r.df <- subset(feat_enrich_agg, MutFreq != 0 & SwitchFreq != 0)
r <- cor(r.df$MutFreq/r.df$DomainFrequency,r.df$SwitchFreq/r.df$DomainFrequency)

p <- ggplot(data=r.df,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
  geom_point() + 
  geom_point(data=subset(r.df,MutFreq/DomainFrequency > minMut & 
                         SwitchFreq/DomainFrequency > minSwitch),
             aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency,color=Domain)) + 
  smartas_theme() + 
  geom_smooth(method=lm) + 
  geom_text(x=6,y=20,label=paste0("R = ",round(r,2)))
p <- direct.label(p)

p
```

Similar results appear, but the correlation is non-existant, suggesting that the domains that need to be affected in each cancer type are tumor specific.

## Clustering by domain function

We wanted to know if the correlation improves if we aggregate the domains by function affected. So, maybe if domains that are only affected by mutations or by switches actually share some functions.

The frequencies are calculated as the quotient between the "in" counts (mutations in the feature, switches affecting the feature, of number of times that feature is observed in the proteome) and the total number of counts.

```{r feat_by_GO_preparation}

# read GO termns
prosite_annotation <- read.delim(paste0(wd,"prosite2go.clean.txt"),row.names = NULL,header=F)
pfam_annotation <- read.delim(paste0(wd,"Pfam2go.clean.txt"),row.names = NULL,header=F)

annotation <- rbind(prosite_annotation,pfam_annotation)
colnames(annotation) <- c("id","GO")

# remove the GO prefix from all of them  
z <- gsub(";GO:[[:digit:]]+", "", annotation$GO)
z <- gsub("^GO:", "", z)
z <- paste0(toupper(substring(z, 1,1)),substring(z, 2))
  
annotation$GO <- z

# get the interpro id for the domains
feat_enrich_agg.go <- feat_enrich_agg
feat_enrich_agg.go$id <- unlist(strsplit(as.character(feat_enrich_agg.go$Domain),"|",fixed = T))[c(T,F)]

# merge the two tables. if a domain has several annotations, if will appear several times
feat_enrich_agg.go <- merge(feat_enrich_agg.go,annotation,all.x=T)

# aggregate by GO term and calculate the frequencies
feat_function_aggregated <- ddply(feat_enrich_agg.go,.(GO),summarise,MutIn=sum(MutIn),AllMuts = sum(AllMuts), SwitchesIn = sum(SwitchesIn), AllSwitches = sum(AllSwitches), DomainCount=sum(DomainCount),AllDomains=sum(AllDomains))

feat_function_aggregated$MutFreq <- feat_function_aggregated$MutIn/feat_function_aggregated$AllMuts
feat_function_aggregated$SwitchFreq <- feat_function_aggregated$SwitchesIn/feat_function_aggregated$AllSwitches
feat_function_aggregated$GOFrequency <- feat_function_aggregated$DomainCount/feat_function_aggregated$AllDomains
```

```{r feat_by_GO_raw}

# plot original data
r <- cor(feat_function_aggregated$MutFreq/feat_function_aggregated$GOFrequency,feat_function_aggregated$SwitchFreq/feat_function_aggregated$GOFrequency)

p <- ggplot(data=feat_function_aggregated,aes(MutFreq/GOFrequency,SwitchFreq/GOFrequency)) + 
  geom_point() + 
  smartas_theme() + 
  geom_smooth(method=lm) + 
  geom_text(x=20,y=20,label=paste0("R = ",round(r,2)))

p
```

Same as before, there seems to be a mutual exclusion between functions affected by mutations and functions affected by alternative splicing.

```{r feat_by_GO_filtered}

r.df <- subset(feat_function_aggregated, SwitchFreq != 0 & MutFreq != 0)
r <- cor(r.df$MutFreq/r.df$GOFrequency,r.df$SwitchFreq/r.df$GOFrequency)

# show the extreme values
minMut <- quantile(r.df$MutFreq/r.df$GOFrequency,0.95)
minSwitch <- quantile(r.df$SwitchFreq/r.df$GOFrequency,0.95)

p <- ggplot(data=r.df,aes(MutFreq/GOFrequency,SwitchFreq/GOFrequency)) + 
  geom_point() + 
  geom_point(data=subset(r.df,MutFreq/GOFrequency > minMut & 
                         SwitchFreq/GOFrequency > minSwitch),
             aes(MutFreq/GOFrequency,SwitchFreq/GOFrequency,color=GO)) +
  smartas_theme() + 
  geom_smooth(method=lm) + 
  geom_text(x=6,y=20,label=paste0("R = ",round(r,2)))
p <- direct.label(p)

p
```

Removing the cases with perfect mutual exclusion only stresses the L-shape of the distribution. Only the urocanate metabolism stands out, same as in previous analyses. This amino acid is part of the degradation of the histidine, and any relation to cancer progression is not evident.

## Repeat, filtering the data

The L-shaped plots, might mean there is a mutual exclusion between features affected by mutations and alternative splicing. However, it may be just an illusion due to the outliers that reduce the resolution. So we make the same plot removing the outliers, outlier meaning points more extreme than quantile 95 in either axis.

```{r tumor_mutVsSwitch_minusOutliers}

for (cancer in cancerTypes){
  
  cat(paste0(cancer,"\n"))
  cancer.feat_enrich <- feat_enrich[feat_enrich$Cancer==cancer,]
  
  # plot normalized mutation frequency and switched frequency
  mutOutliers <- quantile(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,0.95)
  switchOutliers <- quantile(cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency,0.95)
  
  # remove outliers
  cancer.feat_enrich <- cancer.feat_enrich[cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency <= mutOutliers,]
  cancer.feat_enrich <- cancer.feat_enrich[cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency <= switchOutliers,]
  
  # plot raw data
  r <- cor(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,
           cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency)
  
  p <- ggplot(cancer.feat_enrich,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
    geom_point() + 
    smartas_theme() + 
    geom_smooth(method=lm) + 
    geom_text(x=4,y=3,label=paste0("R = ",round(r,2)))
  
  print(p)

}
```

The L-shape seems kept in all the tumor types, reaffirming the mutual exclusion.

# Binomial test between frequency of alterations and frequency of feature

In order to see if there is any feature enriched in mutations in general, as we do for individual switches, we want to check if any kind of domain is frequently mutated/switched more than expected.

## Analysis by tumor type

Below are plotted the with an adjusted p-value in either mutations or switches < 0.05, and a frequency higher than 0 in both axis. Also, tables with the top 10 domains ordered by mutation enrichment p-value, and filtered by adjusted p of enrichment in switches < 0.05.

```{r tumor_domainEnrichment_binomiarTest, results='asis' }

for (cancer in cancerTypes){
  
  cat(paste0(cancer,"\n\n"))
  cancer.feat_enrich <- feat_enrich[feat_enrich$Cancer==cancer,]
  
  # binomial tests to mutation affection and switch affection
  cancer.feat_enrich$p_mut <- apply(cancer.feat_enrich[,c("MutIn","AllMuts","DomainFrequency")],1, function(x){ 
    p <- binom.test(x[1],x[2],x[3],"greater")
    p$p.value
  } )
  cancer.feat_enrich$adjp_mut <- p.adjust(cancer.feat_enrich$p_mut)
  
  cancer.feat_enrich$p_swt <- apply(cancer.feat_enrich[,c("SwitchesIn","AllSwitches","DomainFrequency")],1, function(x){ 
    p <- binom.test(x[1],x[2],x[3],"greater")
    p$p.value
  } )
  cancer.feat_enrich$adjp_swt <- p.adjust(cancer.feat_enrich$p_swt)
  
  # plot domains that are enriched either by mutation affection or by switch affection
  cancer.feat_enrich <- cancer.feat_enrich[cancer.feat_enrich$adjp_swt < 0.05 | cancer.feat_enrich$adjp_mut < 0.05,]
  cancer.feat_enrich <- subset(cancer.feat_enrich, MutFreq != 0 & SwitchFreq != 0)
  
  r <- cor(cancer.feat_enrich$MutFreq/cancer.feat_enrich$DomainFrequency,
         cancer.feat_enrich$SwitchFreq/cancer.feat_enrich$DomainFrequency)
  
  p <- ggplot(cancer.feat_enrich,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
    geom_point() + 
    smartas_theme() + 
    geom_smooth(method=lm) + 
    geom_text(x=4,y=50,label=paste0("R = ",round(r,2)))
  
  print(p)
  cat("\n")
  
  # show top 10 domains affected by mutations
  df <- cancer.feat_enrich[order(cancer.feat_enrich$p_mut),c("Domain","adjp_mut","adjp_swt")]
  df <- df[df$adjp_swt < 0.05,]

  # prepare colums for print
  df$Domain <- gsub("|"," - ",df$Domain,fixed = T)
  df$Domain <- gsub("_"," ",df$Domain,fixed = T)
  
  df$adjp_mut <- format(df$adjp_mut,scientific=TRUE)
  df$adjp_swt <- format(df$adjp_swt,scientific=TRUE)

  print(kable(head(df,n = 10),row.names = F, digits=3))
  cat("\n")
  
}
```

Kinase domains are frequently mutated, but not that frequently switched, except in some cases (PF07714). Also, there are some recurrent domains such as (RAS small GTPase and RHO small GTPase). 

## Analysis aggregating all tumors

Although there seems to be a correlation between tumor type and domain affected, it is interesting to aggregate the results to find out if the important domains are kept.

```{r aggregated_domainEnrichment_binomiarTest}

feat_enrich_agg.enrichment <- feat_enrich_agg

feat_enrich_agg.enrichment$p_mut <- apply(feat_enrich_agg.enrichment[,c("MutIn","AllMuts","DomainFrequency")],1, function(x){ 
  p <- binom.test(x[1],x[2],x[3],"greater")
  p$p.value
} )
feat_enrich_agg.enrichment$adjp_mut <- p.adjust(feat_enrich_agg.enrichment$p_mut)

feat_enrich_agg.enrichment$p_swt <- apply(feat_enrich_agg.enrichment[,c("SwitchesIn","AllSwitches","DomainFrequency")],1, function(x){ 
  p <- binom.test(x[1],x[2],x[3],"greater")
  p$p.value
} )
feat_enrich_agg.enrichment$adjp_swt <- p.adjust(feat_enrich_agg.enrichment$p_swt)

# filter out no significant and only affected by one mechanism
df <- feat_enrich_agg.enrichment[feat_enrich_agg.enrichment$adjp_swt < 0.05 | feat_enrich_agg.enrichment$adjp_mut < 0.05,]
df <- subset(df, MutFreq != 0 & SwitchFreq != 0)
  
r <- cor(df$MutFreq/df$DomainFrequency,df$SwitchFreq/df$DomainFrequency)
  
ggplot(df,aes(MutFreq/DomainFrequency,SwitchFreq/DomainFrequency)) + 
  geom_point() + 
  smartas_theme() + 
  geom_smooth(method=lm) + 
  geom_text(x=4,y=50,label=paste0("R = ",round(r,2)))

# show top 50 domains affected by mutations
df <- feat_enrich_agg.enrichment[order(feat_enrich_agg.enrichment$p_mut),c("Domain","adjp_mut","adjp_swt","MutFreq","SwitchFreq","DomainFrequency")]
df <- df[df$adjp_mut < 0.05 | df$adjp_swt < 0.05,]

# prepare colums for print
df$Domain <- gsub("|"," - ",df$Domain,fixed = T)
df$Domain <- gsub("_"," ",df$Domain,fixed = T)

df$adjp_mut <- format(df$adjp_mut,digits=3)
df$adjp_swt <- format(df$adjp_swt,digits=3)
df$fc_mut <- round(df$MutFreq/df$DomainFrequency,2)
df$fc_swt <- round(df$SwitchFreq/df$DomainFrequency,2)

df <- df[order(-df$fc_mut),]

datatable(df[,c("Domain","adjp_mut","adjp_swt","fc_mut","fc_swt")])

```

# TO-DO

We want to be able to rank domains based on mutations, and to know if that is possible. 

1. The expected frequency of mutation should be the length of a domain divided by the total length of the proteome.